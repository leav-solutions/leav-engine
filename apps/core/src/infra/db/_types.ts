// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
