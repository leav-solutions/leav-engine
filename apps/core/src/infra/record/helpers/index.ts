// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export {default as filterTypes} from './filterTypes';
export {default as getClassifyingFiltersVariableQueryPart} from './getClassifyingFiltersVariableQueryPart';
export {default as getSearchVariableName} from './getSearchVariableName';
export {default as getSearchVariablesQueryPart} from './getSearchVariablesQueryPart';
