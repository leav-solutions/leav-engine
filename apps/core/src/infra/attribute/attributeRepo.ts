// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {IUtils} from 'utils/utils';
import {IList} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IAttribute, IGetCoreAttributesParams} from '../../_types/attribute';
import {IGetCoreEntitiesParams} from '../../_types/shared';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';
import {LIB_ATTRIB_COLLECTION_NAME, LIB_COLLECTION_NAME} from '../library/libraryRepo';
import {IValueRepo} from '../value/valueRepo';

export interface IAttributeRepo {
    getAttributes({params, ctx}: {params?: IGetCoreAttributesParams; ctx: IQueryInfos}): Promise<IList<IAttribute>>;
    updateAttribute({attrData, ctx}: {attrData: IAttributeForRepo; ctx: IQueryInfos}): Promise<IAttribute>;
    createAttribute({attrData, ctx}: {attrData: IAttributeForRepo; ctx: IQueryInfos}): Promise<IAttribute>;
    deleteAttribute({attrData, ctx}: {attrData: IAttribute; ctx: IQueryInfos}): Promise<IAttribute>;
    getLibraryAttributes({libraryId, ctx}: {libraryId: string; ctx: IQueryInfos}): Promise<IAttribute[]>;
    getLibraryFullTextAttributes({libraryId, ctx}: {libraryId: string; ctx: IQueryInfos}): Promise<IAttribute[]>;
}

export const ATTRIB_COLLECTION_NAME = 'core_attributes';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.infra.value'?: IValueRepo;
    'core.utils'?: IUtils;
}

export interface IAttributeForRepo extends IAttribute {
    multiple_values: boolean;
}

export default function({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.infra.value': valueRepo = null,
    'core.utils': utils = null
}: IDeps = {}): IAttributeRepo {
    return {
        async getLibraryAttributes({libraryId, ctx}): Promise<IAttribute[]> {
            // TODO: use aql template tag, and find out why it doesn't work :)
            const query = `
                FOR v
                IN 1 OUTBOUND '${LIB_COLLECTION_NAME}/${libraryId}'
                ${LIB_ATTRIB_COLLECTION_NAME}
                RETURN v
            `;

            const res = await dbService.execute({query, ctx});

            return res.map(a => dbUtils.cleanup<IAttribute>(a));
        },
        async getLibraryFullTextAttributes({libraryId, ctx}): Promise<IAttribute[]> {
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);
            const attributesCollec = dbService.db.edgeCollection(ATTRIB_COLLECTION_NAME);

            const attrs = await dbService.execute({
                query: aql`LET fullTextAttrs = (
                            FOR e IN ${libAttributesCollec}
                                FILTER e._from == ${LIB_COLLECTION_NAME + '/' + libraryId}
                                FILTER e.full_text_search == true
                            RETURN LAST(SPLIT(e._to, '/'))
                        )
                        FOR a IN ${attributesCollec}
                            FILTER POSITION(fullTextAttrs, a._key)
                        RETURN a
                    `,
                ctx
            });

            return attrs.map(a => dbUtils.cleanup<IAttribute>(a));
        },
        async getAttributes({
            params,
            ctx
        }: {
            params?: IGetCoreAttributesParams;
            ctx: IQueryInfos;
        }): Promise<IList<IAttribute>> {
            const _generateLibrariesFilterConds = (filterKey: string, filterVal: string | boolean | string[]) => {
                if (typeof filterVal === 'boolean') {
                    return aql``;
                }

                const libs = utils.forceArray(filterVal);

                const valParts = libs.map(l => aql`v._key == ${l}`);
                const libKeyCond = aql.join(valParts, ' OR ');

                // Check if there is links between given libraries and attribute
                return aql`LENGTH(
                    FOR v IN 1 INBOUND el
                    core_edge_libraries_attributes
                    FILTER ${libKeyCond}
                    RETURN v._key
                ) > 0`;
            };

            const defaultParams: IGetCoreEntitiesParams = {
                filters: null,
                strictFilters: false,
                withCount: false,
                pagination: null,
                sort: null
            };

            const initializedParams = {...defaultParams, ...params};

            return dbUtils.findCoreEntity<IAttribute>({
                ...initializedParams,
                collectionName: ATTRIB_COLLECTION_NAME,
                customFilterConditions: {libraries: _generateLibrariesFilterConds},
                ctx
            });
        },
        async updateAttribute({attrData, ctx}): Promise<IAttribute> {
            const docToInsert = dbUtils.convertToDoc(attrData);

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`UPDATE ${docToInsert} IN ${col} RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(res.pop());
        },
        async createAttribute({attrData, ctx}): Promise<IAttribute> {
            const docToInsert = dbUtils.convertToDoc(attrData);
            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`INSERT ${docToInsert} IN ${col} RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(res.pop());
        },
        async deleteAttribute({attrData, ctx}): Promise<IAttribute> {
            // Delete links library<->attribute
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // Delete all values
            await valueRepo.clearAllValues({attribute: attrData, ctx});

            await dbService.execute({
                query: aql`
                    FOR e IN ${libAttributesCollec}
                        FILTER e._to == ${'core_attributes/' + attrData.id}
                        REMOVE e IN ${libAttributesCollec}
                    `,
                ctx
            });

            // Delete attribute
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);

            const res = await dbService.execute({
                query: aql`REMOVE ${{_key: attrData.id}} IN ${col} RETURN OLD`,
                ctx
            });

            // Return deleted attribute
            return dbUtils.cleanup(res.pop());
        }
    };
}
