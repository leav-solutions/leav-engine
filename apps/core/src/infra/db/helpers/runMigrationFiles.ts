// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {asFunction, AwilixContainer} from 'awilix';
import {Winston} from 'winston';
import {IMigration} from '_types/migration';
import {IQueryInfos} from '_types/queryInfos';
import {IDbService} from '../dbService';
import {MIGRATIONS_COLLECTION_NAME} from '../dbUtils';
import loadMigrationFile from './loadMigrationFile';

interface IExecuteMigrationParams {
    files: string[];
    executedMigrations: string[];
    migrationsDir: string;
    prefix?: string;
    deps?: {
        depsManager: AwilixContainer;
        dbService: IDbService;
        logger: Winston;
    };
    ctx: IQueryInfos;
}

export default async (params: IExecuteMigrationParams): Promise<void> => {
    const {files, executedMigrations, migrationsDir, prefix = null, deps, ctx} = params;

    for (const file of files) {
        // Check if it's been run before
        const fileKey = prefix ? [prefix, file].join('/') : file;
        if (typeof (executedMigrations as []).find(el => el === fileKey) === 'undefined') {
            const importedFile = await loadMigrationFile(migrationsDir + '/' + file);

            if (typeof importedFile.default !== 'function') {
                throw new Error(`[DB Migration Error] ${fileKey}: Migration files' default export must be a function`);
            }

            try {
                deps.logger.info(`[DB Migration] Executing ${fileKey}...`);

                // Run migration
                const migration: IMigration = deps.depsManager.build(asFunction(importedFile.default));
                await migration.run(ctx);

                // Store migration execution to DB
                const collection = deps.dbService.db.collection(MIGRATIONS_COLLECTION_NAME);
                await collection.save({
                    file: fileKey,
                    date: Date.now()
                });
            } catch (err) {
                err.message = `[DB Migration Error] ${fileKey}: } ${err.message}`;
                throw err;
            }
        }
    }
};
