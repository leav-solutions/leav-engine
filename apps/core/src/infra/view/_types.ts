// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
