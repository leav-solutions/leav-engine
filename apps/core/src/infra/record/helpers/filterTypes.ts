// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {AttributeCondition, IRecordFilterOption, TreeCondition} from '../../../_types/record';

export interface IFilterTypesHelper {
    isAttributeFilter(filter: IRecordFilterOption): boolean;
    isClassifyingFilter(filter: IRecordFilterOption): boolean;
    isCountFilter(filter: IRecordFilterOption): boolean;
}

export default function(): IFilterTypesHelper {
    const _isAttributeFilter = (filter: IRecordFilterOption) => {
        return (filter as IRecordFilterOption).condition in AttributeCondition;
    };

    const _isClassifyingFilter = (filter: IRecordFilterOption) => {
        return (filter as IRecordFilterOption).condition in TreeCondition;
    };

    const _isCountFilter = (filter: IRecordFilterOption) => {
        return [
            AttributeCondition.VALUES_COUNT_EQUAL,
            AttributeCondition.VALUES_COUNT_GREATER_THAN,
            AttributeCondition.VALUES_COUNT_LOWER_THAN,
            AttributeCondition.IS_EMPTY,
            AttributeCondition.IS_NOT_EMPTY
        ].includes(filter.condition as AttributeCondition);
    };

    return {
        isAttributeFilter: _isAttributeFilter,
        isClassifyingFilter: _isClassifyingFilter,
        isCountFilter: _isCountFilter
    };
}
