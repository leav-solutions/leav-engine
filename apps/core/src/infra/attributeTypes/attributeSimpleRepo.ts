// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, AqlQuery, GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';
import {IRecordSort} from '_types/record';
import {AttributeFormats, IAttribute} from '../../_types/attribute';
import {IStandardValue, IValue} from '../../_types/value';
import {ATTRIB_COLLECTION_NAME} from '../attribute/attributeRepo';
import {IDbService} from '../db/dbService';
import {LIB_ATTRIB_COLLECTION_NAME} from '../library/libraryRepo';
import {BASE_QUERY_IDENTIFIER, IAttributeTypeRepo} from './attributeTypesRepo';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
}

export default function ({'core.infra.db.dbService': dbService = null}: IDeps = {}): IAttributeTypeRepo {
    async function _saveValue(
        library: string,
        recordId: string,
        attribute: IAttribute,
        value: IValue,
        ctx: IQueryInfos
    ): Promise<IValue> {
        const collec = dbService.db.collection(library);

        const res = await dbService.execute({
            query: aql`
                UPDATE ${{_key: recordId}}
                WITH ${{[attribute.id]: value.value}}
                IN ${collec}
                OPTIONS { keepNull: false }
                RETURN NEW`,
            ctx
        });
        const updatedDoc = res.length ? res[0] : {};

        return {
            value: typeof updatedDoc[attribute.id] !== 'undefined' ? updatedDoc[attribute.id] : null,
            created_by: null,
            modified_by: null
        };
    }

    function _getExtendedFilterPart(attributes: IAttribute[]): GeneratedAqlQuery {
        return aql`${
            attributes
                .map(a => a.id)
                .reduce((acc, value, i) => {
                    acc.push(aql`TRANSLATE(${value}, ${i ? acc[acc.length - 1] : aql`r`})`);
                    if (i) {
                        acc.shift();
                    }
                    return acc;
                }, [])[0]
        }`;
    }

    return {
        async createValue({library, recordId, attribute, value, ctx}): Promise<IStandardValue> {
            return _saveValue(library, recordId, attribute, value, ctx);
        },
        async updateValue({library, recordId, attribute, value, ctx}): Promise<IStandardValue> {
            return _saveValue(library, recordId, attribute, value, ctx);
        },
        async deleteValue({library, recordId, attribute, value, ctx}): Promise<IStandardValue> {
            return _saveValue(library, recordId, attribute, {...value, value: null}, ctx);
        },
        async getValues({library, recordId, attribute, ctx}): Promise<IStandardValue[]> {
            const query = aql`
                FOR r IN ${dbService.db.collection(library)}
                    FILTER r._key == ${String(recordId)}
                    RETURN r.${attribute.id}
            `;
            const res = await dbService.execute({query, ctx});

            return [
                {
                    value: res[0],
                    attribute: attribute.id,
                    modified_by: null,
                    created_by: null
                }
            ];
        },
        async getValueById(): Promise<IStandardValue> {
            return null;
        },
        filterQueryPart(
            attributes: IAttribute[],
            queryPart: GeneratedAqlQuery,
            parentIdentifier = BASE_QUERY_IDENTIFIER
        ): AqlQuery {
            attributes[0].id = attributes[0].id === 'id' ? '_key' : attributes[0].id;

            const query: AqlQuery =
                attributes[0].format === AttributeFormats.EXTENDED && attributes.length > 1
                    ? aql`FILTER ${_getExtendedFilterPart(attributes)} ${queryPart}`
                    : aql`FILTER ${aql.literal(parentIdentifier)}.${attributes[0].id} ${queryPart}`;

            return query;
        },
        sortQueryPart({attributes, order}: IRecordSort): AqlQuery {
            attributes[0].id = attributes[0].id === 'id' ? '_key' : attributes[0].id;

            const query: AqlQuery =
                attributes[0].format === AttributeFormats.EXTENDED && attributes.length > 1
                    ? aql`SORT ${_getExtendedFilterPart(attributes)} ${order}`
                    : aql`SORT r.${attributes[0].id} ${order}`;

            return query;
        },
        async clearAllValues({attribute, ctx}): Promise<boolean> {
            const libAttribCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // TODO: use aql template tag, and find out why it doesn't work :)
            const query = `
                FOR v
                IN 1 INBOUND '${ATTRIB_COLLECTION_NAME}/${attribute.id}'
                ${LIB_ATTRIB_COLLECTION_NAME}
                RETURN v
            `;

            const libraries = await dbService.execute({query, ctx});

            for (const lib of libraries) {
                const recordsCollec = dbService.db.collection(lib._key);
                const clearQuery = aql`
                    FOR r IN ${recordsCollec}
                    FILTER r.${attribute.id} != null
                    UPDATE r WITH {${attribute.id}: null} IN ${recordsCollec} OPTIONS {keepNull: false}
                `;

                await dbService.execute({query: clearQuery, ctx});
            }

            return true;
        }
    };
}
