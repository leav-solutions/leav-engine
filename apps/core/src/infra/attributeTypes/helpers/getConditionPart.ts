// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, AqlLiteral, GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import moment from 'moment';
import {AttributeFormats, IAttribute} from '../../../_types/attribute';
import {AttributeCondition, IDateFilterValue} from '../../../_types/record';

export type GetConditionPart = (
    valueIdentifier: string | AqlLiteral,
    condition: AttributeCondition,
    value: string | number | boolean,
    attribute: IAttribute
) => GeneratedAqlQuery;

export default function (): GetConditionPart {
    return (
        valueIdentifier: string | AqlLiteral,
        condition: AttributeCondition,
        value: string | number | boolean | IDateFilterValue,
        attribute: IAttribute
    ): GeneratedAqlQuery => {
        const valueField = typeof valueIdentifier === 'string' ? aql.literal(valueIdentifier) : valueIdentifier;

        switch (condition) {
            case AttributeCondition.EQUAL: {
                const cond =
                    attribute.format === AttributeFormats.DATE
                        ? aql`DATE_COMPARE(${valueField} * 1000, ${Number(value) * 1000}, "years", "days") == true`
                        : aql`${valueField} == ${value}`;

                return cond;
            }
            case AttributeCondition.NOT_EQUAL:
                return attribute.format === AttributeFormats.DATE
                    ? aql`DATE_COMPARE(${valueField} * 1000, ${Number(value) * 1000}, "years", "days") == false`
                    : aql`${valueField} != ${value}`;
            case AttributeCondition.BEGIN_WITH:
                return aql`${valueField} LIKE ${`${value}%`}`;
            case AttributeCondition.END_WITH:
                return aql`${valueField} LIKE ${`%${value}`}`;
            case AttributeCondition.CONTAINS: {
                return attribute.format === AttributeFormats.DATE_RANGE
                    ? aql`(${Number(value)} >= ${valueField}.from AND ${Number(value)} <= ${valueField}.to)`
                    : aql`${valueField} LIKE ${`%${value}%`}`;
            }
            case AttributeCondition.NOT_CONTAINS:
                return aql`${valueField} NOT LIKE ${`%${value}%`}`;
            case AttributeCondition.GREATER_THAN:
                return aql`${valueField} > ${Number(value)}`;
            case AttributeCondition.LESS_THAN:
                return aql`(${valueField} != null AND ${valueField} < ${Number(value)})`;
            case AttributeCondition.IS_EMPTY:
                return aql`${valueField} == null`;
            case AttributeCondition.IS_NOT_EMPTY:
                return aql`${valueField} != null`;
            case AttributeCondition.BETWEEN:
                return aql`(
                            ${valueField} >= ${Number((value as IDateFilterValue).from)}
                            AND ${valueField} <= ${Number((value as IDateFilterValue).to)}
                        )`;
            case AttributeCondition.TODAY:
                return aql`DATE_COMPARE(${valueField} * 1000, DATE_NOW(), "years", "days") == true`;
            case AttributeCondition.YESTERDAY:
                return aql`DATE_COMPARE(
                            ${valueField} * 1000,
                            DATE_SUBTRACT(DATE_NOW(), 1, "day"),
                            "years",
                            "days"
                        ) == true`;
            case AttributeCondition.TOMORROW:
                return aql`DATE_COMPARE(
                            ${valueField} * 1000,
                            DATE_ADD(DATE_NOW(), 1, "day"),
                            "years",
                            "days"
                        ) == true`;
            case AttributeCondition.NEXT_MONTH: {
                const now = moment().unix();
                const nextMonth = moment().add(31, 'days').unix();
                return aql`(${valueField} >= ${now} AND ${valueField} <=${nextMonth})`;
            }
            case AttributeCondition.LAST_MONTH: {
                const now = moment().unix();
                const lastMonth = moment().subtract(31, 'days').unix();
                return aql`(${valueField} >= ${lastMonth} AND ${valueField} <=${now})`;
            }
            case AttributeCondition.START_ON:
                return aql`DATE_COMPARE(${valueField}.from * 1000, ${Number(value) * 1000}, "years", "days") == true`;
            case AttributeCondition.START_BEFORE:
                return aql`(${valueField}.from != null AND ${valueField}.from < ${Number(value)})`;
            case AttributeCondition.START_AFTER:
                return aql`${valueField}.from > ${Number(value)}`;
            case AttributeCondition.END_ON:
                return aql`DATE_COMPARE(${valueField}.to * 1000, ${Number(value) * 1000}, "years", "days") == true`;
            case AttributeCondition.END_BEFORE:
                return aql`(${valueField}.to != null AND ${valueField}.to < ${Number(value)})`;
            case AttributeCondition.END_AFTER:
                return aql`${valueField}.to > ${Number(value)}`;
            default:
                return aql`${valueField} == ${value}`;
        }
    };
}
