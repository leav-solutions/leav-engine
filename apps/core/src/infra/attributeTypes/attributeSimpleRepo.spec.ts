// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IFilterTypesHelper} from 'infra/record/helpers/filterTypes';
import {IQueryInfos} from '_types/queryInfos';
import {AttributeFormats, AttributeTypes} from '../../_types/attribute';
import {AttributeCondition} from '../../_types/record';
import {mockAttrSimple} from '../../__tests__/mocks/attribute';
import attributeSimpleRepo from './attributeSimpleRepo';

describe('AttributeSimpleRepo', () => {
    const mockAttribute = {
        id: 'test_attr',
        type: AttributeTypes.SIMPLE
    };
    const ctx: IQueryInfos = {
        userId: '0',
        queryId: 'attributeSimpleRepoTest'
    };

    describe('createValue', () => {
        test('Should create a new index value', async function() {
            const updatedRecordData = {
                _id: 'test_lib/222435651',
                _rev: '_WSywvyC--_',
                _key: 222435651,
                test_attr: 'test_val'
            };

            const updatedValueData = {
                value: 'test_val'
            };

            const mockDbServ = {db: new Database(), execute: global.__mockPromise([updatedRecordData])};

            const attrRepo = attributeSimpleRepo({'core.infra.db.dbService': mockDbServ});

            const createdVal = await attrRepo.createValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 'test val'
                },
                ctx
            });

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(createdVal).toMatchObject(updatedValueData);
        });
    });

    describe('getValues', () => {
        test('Should return values for index attribute', async function() {
            const queryRes = ['test val', 'other val that should not be returned'];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(queryRes)
            };

            const attrRepo = attributeSimpleRepo({'core.infra.db.dbService': mockDbServ});

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttribute,
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(values.length).toBe(1);
            expect(values[0]).toMatchObject({
                value: 'test val',
                attribute: 'test_attr'
            });
        });
    });

    describe('deleteValue', () => {
        test('Should delete a value', async () => {
            const updatedRecordData = {
                _id: 'test_lib/222435651',
                _rev: '_WSywvyC--_',
                _key: 222435651
            };

            const deletedValueData = {
                value: null
            };

            const mockDbServ = {db: new Database(), execute: global.__mockPromise([updatedRecordData])};

            const attrRepo = attributeSimpleRepo({'core.infra.db.dbService': mockDbServ});

            const deletedVal = await attrRepo.deleteValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 'test val'
                },
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(deletedVal).toMatchObject(deletedValueData);
        });
    });

    describe('sortQueryPart', () => {
        test('Should return simple sort', () => {
            const attrRepo = attributeSimpleRepo();
            const filter = attrRepo.sortQueryPart({
                attributes: [{id: 'id', type: AttributeTypes.SIMPLE}],
                order: 'ASC'
            });

            expect(filter.query).toMatch(/^SORT/);
            expect(filter).toMatchSnapshot();
        });
    });

    describe('filterValueQueryPart', () => {
        const mockFilterTypesHelper: Mockify<IFilterTypesHelper> = {
            isCountFilter: jest.fn().mockReturnValue(false)
        };

        test('Should query to retrieve value to filter on', () => {
            const attrRepo = attributeSimpleRepo({
                'core.infra.record.helpers.filterTypes': mockFilterTypesHelper as IFilterTypesHelper
            });
            const valueQuery = attrRepo.filterValueQueryPart([{...mockAttrSimple, _repo: null, reverse_link: null}], {
                condition: AttributeCondition.EQUAL,
                attributes: [{...mockAttrSimple, reverse_link: null}],
                value: '123456'
            });

            expect(valueQuery.query).toMatch(/^r\./);
            expect(valueQuery).toMatchSnapshot();
        });

        test('Should query to retrieve value to filter on, on extended format', () => {
            const attrRepo = attributeSimpleRepo({
                'core.infra.record.helpers.filterTypes': mockFilterTypesHelper as IFilterTypesHelper
            });
            const valueQuery = attrRepo.filterValueQueryPart(
                [
                    {...mockAttrSimple, format: AttributeFormats.EXTENDED, _repo: null, reverse_link: null},
                    {...mockAttrSimple, id: 'subfield', _repo: null, reverse_link: null}
                ],
                {
                    condition: AttributeCondition.EQUAL,
                    attributes: [{...mockAttrSimple, reverse_link: null}],
                    value: '123456'
                }
            );

            expect(valueQuery).toMatchSnapshot();
        });
    });

    describe('clearAllValues', () => {
        test('Should delete field of given attribute', async () => {
            const librarieusUsingAttribute = [
                {
                    _key: 'users',
                    label: {en: 'Users', fr: 'Utilisateurs'},
                    system: true
                },
                {
                    _key: 'products',
                    label: {en: 'Products', fr: 'Produits'},
                    system: false
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: jest
                    .fn()
                    .mockReturnValueOnce(librarieusUsingAttribute) // Number of deleted values
                    .mockReturnValueOnce([]) // Number of deleted values
                    .mockReturnValueOnce([])
            };

            const attrTypeRepo = attributeSimpleRepo({'core.infra.db.dbService': mockDbServ});
            const res = await attrTypeRepo.clearAllValues({
                attribute: {id: 'test_attr', type: AttributeTypes.SIMPLE},
                ctx
            });

            expect(res).toEqual(true);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });
    });
});
