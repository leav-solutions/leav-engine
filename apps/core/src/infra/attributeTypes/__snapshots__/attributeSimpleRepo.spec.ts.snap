// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeIndexRepo clearAllValues Should delete field of given attribute 1`] = `undefined`;

exports[`AttributeIndexRepo clearAllValues Should delete field of given attribute 2`] = `undefined`;

exports[`AttributeIndexRepo createValue Should create a new index value 1`] = `
"
                UPDATE @value0
                WITH @value1
                IN @@value2
                OPTIONS { keepNull: false }
                RETURN NEW"
`;

exports[`AttributeIndexRepo createValue Should create a new index value 2`] = `
{
  "@value2": "test_lib",
  "value0": {
    "_key": "12345",
  },
  "value1": {
    "test_attr": "test val",
  },
}
`;

exports[`AttributeIndexRepo deleteValue Should delete a value 1`] = `
"
                UPDATE @value0
                WITH @value1
                IN @@value2
                OPTIONS { keepNull: false }
                RETURN NEW"
`;

exports[`AttributeIndexRepo deleteValue Should delete a value 2`] = `
{
  "@value2": "test_lib",
  "value0": {
    "_key": "12345",
  },
  "value1": {
    "test_attr": null,
  },
}
`;

exports[`AttributeIndexRepo filterQueryPart Should return simple filter 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "value0": "_key",
    "value1": "123456",
  },
  "query": "
                        LET rVal = r.@value0
                        FILTER rVal == @value1",
}
`;

exports[`AttributeIndexRepo getValues Should return values for index attribute 1`] = `
"
                FOR r IN @@value0
                    FILTER r._key == @value1
                    RETURN r.@value2
            "
`;

exports[`AttributeIndexRepo getValues Should return values for index attribute 2`] = `
{
  "@value0": "test_lib",
  "value1": "123456",
  "value2": "test_attr",
}
`;

exports[`AttributeIndexRepo sortQueryPart Should return simple sort 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "value0": "_key",
    "value1": "ASC",
  },
  "query": "SORT r.@value0 @value1",
}
`;
